torchmultimodal/modules/layers/quantization.py

```diff
@@ -117,36 +140,34 @@ def _ema_update_embedding(self, encoded_flat: Tensor, codebook_indices: Tensor):
         # Count how often each embedding vector was looked up
         codebook_selection_count = torch.sum(codebook_onehot, 0)
         # Update usage value for each embedding vector
-        self._code_usage = self._code_usage * self._decay + codebook_selection_count * (
+        self.code_usage = self.code_usage * self._decay + codebook_selection_count * (
             1 - self._decay
         )
         # Laplace smoothing of codebook usage - to prevent zero counts
-        n = torch.sum(self._code_usage)
-        self._code_usage = (
-            (self._code_usage + self._epsilon)
+        n = torch.sum(self.code_usage)
+        self.code_usage = (
+            (self.code_usage + self._epsilon)
             / (n + self.num_embeddings * self._epsilon)
             * n
         )
         # Get all encoded vectors attracted to each embedding vector
         encoded_per_codebook = torch.matmul(codebook_onehot.t(), encoded_flat)
         # Update each embedding vector with new encoded vectors that are attracted to it,
         # divided by its usage to yield the mean of encoded vectors that choose it
-        self._code_avg = (
-            self._code_avg * self._decay + (1 - self._decay) * encoded_per_codebook
-        )
-        self.embedding.weight = nn.Parameter(
-            self._code_avg / self._code_usage.unsqueeze(1)
+        self.code_avg = (
+            self.code_avg * self._decay + (1 - self._decay) * encoded_per_codebook
         )
+        self.embedding = self.code_avg / self.code_usage.unsqueeze(1)
 
     def _quantize(self, encoded_flat: Tensor) -> Tuple[Tensor, Tensor]:
         # Calculate distances from each encoder, E(x), output vector to each embedding vector, e, ||E(x) - e||^2
-        distances = torch.cdist(encoded_flat, self.embedding.weight, p=2.0) ** 2
+        distances = torch.cdist(encoded_flat, self.embedding, p=2.0) ** 2
 
         # Encoding - select closest embedding vectors
         codebook_indices = torch.argmin(distances, dim=1)
 
         # Quantize
-        quantized_flat = self.embedding(codebook_indices)
+        quantized_flat = F.embedding(codebook_indices, self.embedding)
```

Comment:
Is `F.embedding` better than simply indexing `self.embedding`?

Reply:
self.embedding is no longer an `nn.Embedding`, just a Tensor initialized by `register_buffer`